pkg/kkrtpsi/receiver.go

```diff
@@ -0,0 +1,164 @@
+// use rw as the communication layer
+func NewReceiver(rw io.ReadWriter) *Receiver {
+	return &Receiver{rw: rw}
+}
+
+// Intersect on matchables read from the identifiers channel,
+// returning the matching intersection, using the KKRTPSI protocol.
+// The format of an indentifier is string
+// example:
+//  0e1f461bbefa6e07cc2ef06b9ee1ed25101e24d4345af266ed2f5a58bcd26c5e
+func (r *Receiver) Intersect(ctx context.Context, n int64, identifiers <-chan []byte) (intersection [][]byte, err error) {
+	// fetch and set up logger
+	logger := logr.FromContextOrDiscard(ctx)
+	logger = logger.WithValues("protocol", "kkrtpsi")
+
+	// start timer:
+	start := time.Now()
+	timer := time.Now()
+	var mem uint64
+
+	var seeds [cuckoo.Nhash][]byte
+	var oprfOutput = make([]map[uint64]uint64, cuckoo.Nhash)
+	var cuckooHashTable *cuckoo.Cuckoo
+	var sk []byte
+
+	// stage 1: read the hash seeds from the remote side
+	//          initiate a cuckoo hash table and insert all local
+	//          IDs into the cuckoo hash table.
+	stage1 := func() error {
+		logger.V(1).Info("Starting stage 1")
+		for i := range seeds {
+			seeds[i] = make([]byte, hash.SaltLength)
+			if _, err := io.ReadFull(r.rw, seeds[i]); err != nil {
+				return fmt.Errorf("stage1: %v", err)
+			}
+		}
+
+		// send size
+		if err := binary.Write(r.rw, binary.BigEndian, &n); err != nil {
+			return err
+		}
+
+		// instantiate cuckoo hash table
+		cuckooHashTable, err = cuckoo.NewCuckoo(uint64(n), seeds)
+		if err != nil {
+			return err
+		}
+
+		for id := range identifiers {
+			if err = cuckooHashTable.Insert(id); err != nil {
+				return err
+			}
+		}
+
+		// receive secret key for AES-128 (16 byte)
+		// use the first seed as the 32-byte key for highway hashing
+		sk = make([]byte, 16)
+		if _, err := io.ReadFull(r.rw, sk); err != nil {
+			return fmt.Errorf("stage1: %v", err)
+		}
+
+		// end stage1
+		timer, mem = printStageStats(logger, 1, start, start, 0)
+		logger.V(1).Info("Finished stage 1")
+		return nil
+	}
+
+	// stage 2: prepare OPRF receive input and run Receive to get OPRF output
+	stage2 := func() error {
+		logger.V(1).Info("Starting stage 2")
+		oprfInputSize := int(cuckooHashTable.Len())
+		oprfOutput, err = oprf.NewOPRF(oprfInputSize).Receive(cuckooHashTable, sk, r.rw)
+		if err != nil {
+			return err
+		}
+
+		// end stage2
+		timer, mem = printStageStats(logger, 2, timer, start, mem)
+		logger.V(1).Info("Finished stage 2")
+		return nil
+	}
+
+	// stage 3: read remote encoded identifiers and compare
+	//          to produce intersections
+	stage3 := func() error {
+		logger.V(1).Info("Starting stage 3")
+		// read number of remote IDs
+		var remoteN int64
+		if err := binary.Read(r.rw, binary.BigEndian, &remoteN); err != nil {
+			return err
+		}
+
+		// Add a buffer of 64k to amortize syscalls cost
+		var bufferedReader = bufio.NewReaderSize(r.rw, 1024*64)
+
+		// read remote encodings and intersect
+		var remoteEncodings [cuckoo.Nhash]uint64
```

Comment:
You can declare this variable in the loop body, it won't result in additional cost.